GPU circuit breaker OPEN: cuda_error - Detected mismatch between collectives on ranks

**使用deepseed zero2 8卡训练 flux2 full fintune, 经常报多卡通信问题，报错如下：**
[RANK 0] 2026-05-19 10:10:29,299 [INFO] GPU circuit breaker initialized with 8 GPU(s)
[RANK 4] 2026-05-19 10:10:29,604 [ERROR] GPU circuit breaker OPEN: cuda_error - Detected mismatch between collectives on ranks. Rank 4 is running collective: CollectiveFingerPrint(SequenceNumber=101, OpType=ALLGATHER, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))), but Rank 7 is running collective: CollectiveFingerPrint(SequenceNumber=0OpType=REDUCE).Collectives differ in the following aspects:   Sequence number: 101vs 0  Op type: ALLGATHERvs REDUCE  Tensor Tensor shapes: 1vs   Tensor Tensor dtypes: Longvs   Tensor Tensor devices: TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))vs  (GPU 4)
[RANK 6] 2026-05-19 10:10:29,659 [ERROR] GPU circuit breaker OPEN: cuda_error - Detected mismatch between collectives on ranks. Rank 6 is running collective: CollectiveFingerPrint(SequenceNumber=101, OpType=ALLGATHER, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))), but Rank 7 is running collective: CollectiveFingerPrint(SequenceNumber=0OpType=REDUCE).Collectives differ in the following aspects:   Sequence number: 101vs 0  Op type: ALLGATHERvs REDUCE  Tensor Tensor shapes: 1vs   Tensor Tensor dtypes: Longvs   Tensor Tensor devices: TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))vs  (GPU 6)
[RANK 0] 2026-05-19 10:10:29,661 [ERROR] GPU circuit breaker OPEN: cuda_error - Detected mismatch between collectives on ranks. Rank 0 is running collective: CollectiveFingerPrint(SequenceNumber=101, OpType=ALLGATHER, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))), but Rank 7 is running collective: CollectiveFingerPrint(SequenceNumber=0OpType=REDUCE).Collectives differ in the following aspects:   Sequence number: 101vs 0  Op type: ALLGATHERvs REDUCE  Tensor Tensor shapes: 1vs   Tensor Tensor dtypes: Longvs   Tensor Tensor devices: TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))vs  (GPU 0)
[RANK 0] 2026-05-19 10:10:29,663 [INFO] Unloading all model components...
[RANK 5] 2026-05-19 10:10:29,672 [ERROR] GPU circuit breaker OPEN: cuda_error - Detected mismatch between collectives on ranks. Rank 5 is running collective: CollectiveFingerPrint(SequenceNumber=101, OpType=ALLGATHER, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))), but Rank 7 is running collective: CollectiveFingerPrint(SequenceNumber=0OpType=REDUCE).Collectives differ in the following aspects:   Sequence number: 101vs 0  Op type: ALLGATHERvs REDUCE  Tensor Tensor shapes: 1vs   Tensor Tensor dtypes: Longvs   Tensor Tensor devices: TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))vs  (GPU 5)
[RANK 3] 2026-05-19 10:10:29,674 [ERROR] GPU circuit breaker OPEN: cuda_error - Detected mismatch between collectives on ranks. Rank 3 is running collective: CollectiveFingerPrint(SequenceNumber=101, OpType=ALLGATHER, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))), but Rank 7 is running collective: CollectiveFingerPrint(SequenceNumber=0OpType=REDUCE).Collectives differ in the following aspects:   Sequence number: 101vs 0  Op type: ALLGATHERvs REDUCE  Tensor Tensor shapes: 1vs   Tensor Tensor dtypes: Longvs   Tensor Tensor devices: TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))vs  (GPU 3)
[RANK 2] 2026-05-19 10:10:29,692 [ERROR] GPU circuit breaker OPEN: cuda_error - Detected mismatch between collectives on ranks. Rank 2 is running collective: CollectiveFingerPrint(SequenceNumber=101, OpType=ALLGATHER, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))), but Rank 7 is running collective: CollectiveFingerPrint(SequenceNumber=0OpType=REDUCE).Collectives differ in the following aspects:   Sequence number: 101vs 0  Op type: ALLGATHERvs REDUCE  Tensor Tensor shapes: 1vs   Tensor Tensor dtypes: Longvs   Tensor Tensor devices: TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))vs  (GPU 2)
[RANK 1] 2026-05-19 10:10:29,701 [ERROR] GPU circuit breaker OPEN: cuda_error - Detected mismatch between collectives on ranks. Rank 1 is running collective: CollectiveFingerPrint(SequenceNumber=101, OpType=ALLGATHER, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))), but Rank 7 is running collective: CollectiveFingerPrint(SequenceNumber=0OpType=REDUCE).Collectives differ in the following aspects:   Sequence number: 101vs 0  Op type: ALLGATHERvs REDUCE  Tensor Tensor shapes: 1vs   Tensor Tensor dtypes: Longvs   Tensor Tensor devices: TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))vs  (GPU 1)
[RANK 0] 2026-05-19 10:10:30,112 [INFO] Model components unloaded successfully.
[rank7]: Traceback (most recent call last):
[rank7]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/simpletuner/train.py", line 225, in <module>
[rank7]:     trainer.train()
[rank7]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/simpletuner/helpers/training/trainer.py", line 6255, in train
[rank7]:     checkpoint_dir = self._run_standard_checkpoint(
[rank7]:                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank7]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/simpletuner/helpers/training/trainer.py", line 4667, in _run_standard_checkpoint
[rank7]:     self.accelerator.wait_for_everyone()
[rank7]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/simpletuner/helpers/training/trainer.py", line 1043, in _safe_wait_for_everyone
[rank7]:     return original_wait_for_everyone(*fn_args, **fn_kwargs)
[rank7]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank7]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/accelerate/accelerator.py", line 3256, in wait_for_everyone
[rank7]:     wait_for_everyone()
[rank7]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/accelerate/utils/other.py", line 320, in wait_for_everyone
[rank7]:     PartialState().wait_for_everyone()
[rank7]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/accelerate/state.py", line 412, in wait_for_everyone
[rank7]:     torch.distributed.barrier(device_ids=[self.local_process_index])
[rank7]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/torch/distributed/c10d_logger.py", line 83, in wrapper
[rank7]:     return func(*args, **kwargs)
[rank7]:            ^^^^^^^^^^^^^^^^^^^^^
[rank7]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/torch/distributed/distributed_c10d.py", line 5230, in barrier
[rank7]:     work = group.barrier(opts=opts)
[rank7]:            ^^^^^^^^^^^^^^^^^^^^^^^^
[rank7]: RuntimeError: Detected mismatch between collectives on ranks. Rank 7 is running collective: CollectiveFingerPrint(SequenceNumber=101OpType=BARRIER), but Rank 0 is running collective: CollectiveFingerPrint(SequenceNumber=0OpType=GATHER).Collectives differ in the following aspects:     Sequence number: 101vs 0  Op type: BARRIERvs GATHER
[rank0]: Traceback (most recent call last):
[rank0]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/simpletuner/train.py", line 225, in <module>
[rank0]:     trainer.train()
[rank0]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/simpletuner/helpers/training/trainer.py", line 5767, in train
[rank0]:     avg_loss = self.accelerator.gather(loss.repeat(self.config.train_batch_size)).mean()
[rank0]:                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/accelerate/accelerator.py", line 3054, in gather
[rank0]:     return gather(tensor)
[rank0]:            ^^^^^^^^^^^^^^
[rank0]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/accelerate/utils/operations.py", line 384, in wrapper
[rank0]:     output = gather_object([shapes])
[rank0]:              ^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/accelerate/utils/operations.py", line 459, in gather_object
[rank0]:     return _gpu_gather_object(object)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/accelerate/utils/operations.py", line 440, in _gpu_gather_object
[rank0]:     torch.distributed.all_gather_object(output_objects, object)
[rank0]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/torch/distributed/c10d_logger.py", line 83, in wrapper
[rank0]:     return func(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/torch/distributed/distributed_c10d.py", line 3442, in all_gather_object
[rank0]:     all_gather(object_size_list, local_size, group=group)
[rank0]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/torch/distributed/c10d_logger.py", line 83, in wrapper
[rank0]:     return func(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/root/miniconda3/envs/simpletuner_cuda/lib/python3.12/site-packages/torch/distributed/distributed_c10d.py", line 4209, in all_gather
[rank0]:     work = group.allgather([tensor_list], [tensor], opts)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]: RuntimeError: Detected mismatch between collectives on ranks. Rank 0 is running collective: CollectiveFingerPrint(SequenceNumber=101, OpType=ALLGATHER, TensorShape=[1], TensorDtypes=Long, TensorDeviceTypes=TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))), but Rank 7 is running collective: CollectiveFingerPrint(SequenceNumber=0OpType=REDUCE).Collectives differ in the following aspects:    Sequence number: 101vs 0  Op type: ALLGATHERvs REDUCE  Tensor Tensor shapes: 1vs   Tensor Tensor dtypes: Longvs   Tensor Tensor devices: TensorOptions(dtype=float (default), device=cuda, layout=Strided (default), requires_grad=false (default), pinned_memory=false (default), memory_format=(nullopt))vs
[rank0]:I0519 10:10:30.490000 437110 helpers/logging.py:425] TorchDynamo attempted to trace the following frames: [
Show less
**相关的config.json**
{
  "checkpoint_epoch_interval": 1,
  "data_backend_config": "/data1/ywj/SimpleTuner/config/flux2-full-finetune-BaiLiv2.4/multidatabackend.json",
  "caption_shuffle": {
    "enable": true,
    "count": 1,
    "seed": 42,
    "split_on": "comma",
    "position_start": 1,
    "include_original": true
  },
  "mixed_precision":"bf16",
  "disable_multiline_split": false,
  "evaluation_type": "none",
  "flow_schedule_auto_shift": true,
  "gradient_checkpointing": true,
  "learning_rate": 0.0001,
  "lr_scheduler": "constant_with_warmup",
  "lr_warmup_steps": 50,
  "lycoris_config": "/data1/ywj/SimpleTuner/config/lycoris_config.json",
  "max_train_steps": 0,
  "model_family": "flux2",
  "model_flavour": "klein-9b",
  "model_type": "full",
  "num_train_epochs": 5,
  "optimizer": "optimi-adamw",
  "output_dir": "/data1/ywj/SimpleTuner/output/flux2-full-finetune-BaiLiv2.4-dapai",
  "prediction_type": "",
  "pretrained_model_name_or_path": "/data1/ywj/modelzoo/black-forest-labs/FLUX.2-klein-base-9B",
  "publishing_config": "",
  "report_to": "none",
  "scheduled_sampling_reflexflow": false,
  "seed": 42,
  "train_batch_size": 1,
  "gradient_accumulation_steps": 64,
  "use_ema": false,
  "validation_noise_scheduler": "euler",
  "validation_prompt": "A sleek black leather loafer with a polished finish and a subtle pebbled texture, featuring a square toe and a low, chunky heel that provides a balanced, elegant silhouette. The shoe showcases a slip-on design with a slight fold at the vamp, emphasizing its minimalist yet sophisticated style. The interior lining appears to be a rich brown, contrasting with the exterior, while the sole is a matching black with a slight tread pattern, ensuring both comfort and functionality. Set against a clean white background with soft, diffused lighting that highlights the shoe's refined details and smooth leather surface, the overall presentation is crisp and professional. Ultra-high resolution, 8K, sharp focus.",
  "validation_randomize": true,
  "validation_resolution": "1024",
  "validation_step_interval": 10000,
  "validation_using_datasets": false,
  "grad_clip_method":"norm"

}

--- a/simpletuner/helpers/training/trainer.py
+++ b/simpletuner/helpers/training/trainer.py
@@ -345,6 +345,24 @@ class Trainer:
         except Exception:
             return False
 
+    def _distributed_collectives_ready(self) -> bool:
+        try:
+            num_processes = int(getattr(self.accelerator, "num_processes", 1) or 1)
+        except (TypeError, ValueError):
+            num_processes = 1
+        return num_processes > 1 and torch.distributed.is_available() and torch.distributed.is_initialized()
+
+    def _any_rank_reached_epoch_end(self, reached_epoch_end: bool) -> bool:
+        if not self._distributed_collectives_ready():
+            return reached_epoch_end
+
+        device = getattr(self.accelerator, "device", None)
+        if device is None:
+            device = torch.device("cuda", torch.cuda.current_device()) if torch.cuda.is_available() else torch.device("cpu")
+        local_flag = torch.tensor(1 if reached_epoch_end else 0, device=device, dtype=torch.int32)
+        torch.distributed.all_reduce(local_flag, op=torch.distributed.ReduceOp.MAX)
+        return bool(local_flag.item())
+
     def _register_optimizer_attention_params(self, optimizer) -> None:
         try:
             from simpletuner.helpers.training.optimizers.muon import MuonClip
@@ -5609,8 +5627,12 @@ class Trainer:
                         f"Epoch {self.state['current_epoch']}/{self.config.num_train_epochs}, Steps"
                     )
 
-                # If we receive a False from the enumerator, we know we reached the next epoch.
-                if prepared_batch is False:
+                local_epoch_end = prepared_batch is False
+                epoch_end_reached = self._any_rank_reached_epoch_end(local_epoch_end)
+                # If any rank receives False from the enumerator, all ranks must leave the step loop together.
+                if epoch_end_reached:
+                    if not local_epoch_end:
+                        training_logger.debug("Reached the end of epoch because another rank exhausted its dataloader.")
                     if (
                         self._should_checkpoint_epoch(epoch)
                         and not last_step_saved_checkpoint

--- a/tests/test_trainer.py
+++ b/tests/test_trainer.py
@@ -617,6 +617,39 @@ class TestTrainer(unittest.TestCase):
         trainer.grad_norm = grad_value
         return trainer
 
+    def test_any_rank_reached_epoch_end_returns_local_value_without_process_group(self):
+        trainer = object.__new__(Trainer)
+        trainer.accelerator = SimpleNamespace(num_processes=8, device=torch.device("cpu"))
+
+        with (
+            patch("simpletuner.helpers.training.trainer.torch.distributed.is_available", return_value=True),
+            patch("simpletuner.helpers.training.trainer.torch.distributed.is_initialized", return_value=False),
+            patch("simpletuner.helpers.training.trainer.torch.distributed.all_reduce") as mock_all_reduce,
+        ):
+            self.assertTrue(trainer._any_rank_reached_epoch_end(True))
+            self.assertFalse(trainer._any_rank_reached_epoch_end(False))
+
+        mock_all_reduce.assert_not_called()
+
+    def test_any_rank_reached_epoch_end_uses_max_across_distributed_ranks(self):
+        trainer = object.__new__(Trainer)
+        trainer.accelerator = SimpleNamespace(num_processes=8, device=torch.device("cpu"))
+
+        def mark_remote_epoch_end(tensor, op=None):
+            tensor.fill_(1)
+
+        with (
+            patch("simpletuner.helpers.training.trainer.torch.distributed.is_available", return_value=True),
+            patch("simpletuner.helpers.training.trainer.torch.distributed.is_initialized", return_value=True),
+            patch(
+                "simpletuner.helpers.training.trainer.torch.distributed.all_reduce",
+                side_effect=mark_remote_epoch_end,
+            ) as mock_all_reduce,
+        ):
+            self.assertTrue(trainer._any_rank_reached_epoch_end(False))
+
+        mock_all_reduce.assert_called_once()
+
     @patch("simpletuner.helpers.training.trainer.load_config")
     @patch("simpletuner.helpers.training.trainer.safety_check")
     @patch("simpletuner.helpers.training.state_tracker.StateTracker")

